src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+      ldpp_dout(dpp, 1) << "failed to initialize pool for listing with: "
+			<< cpp_strerror(ret) << dendl;
+    }
+
+    return ret;
+  }
+
+  //==========================================================================
+
+  //---------------------------------------------------------------------------
+  void cluster::clear()
+  {
+    d_curr_md5_shard = 0;
+    d_curr_worker_shard = 0;
+
+    d_num_completed_workers = 0;
+    d_num_completed_md5 = 0;
+
+    memset(d_completed_workers, TOKEN_STATE_PENDING, sizeof(d_completed_workers));
+    memset(d_completed_md5, TOKEN_STATE_PENDING, sizeof(d_completed_md5));
+
+    d_total_ingressed_obj = 0;
+    d_num_failed_workers = 0;
+  }
+
+
+  static constexpr auto COOKIE_LEN = 15;
+  static constexpr auto CLUSTER_ID_LEN = 15;
+  //---------------------------------------------------------------------------
+  cluster::cluster(const DoutPrefixProvider *_dpp,
+		   CephContext *cct,
+		   rgw::sal::Driver* driver):
+    dpp(_dpp),
+    d_lock_cookie(gen_rand_alphanumeric(cct, COOKIE_LEN)),
+    d_cluster_id (gen_rand_alphanumeric(cct, CLUSTER_ID_LEN))
+  {
+    clear();
+    ldpp_dout(dpp, 10) << __func__ << "::cluser_id=" << d_cluster_id << dendl;
+
+    auto store = dynamic_cast<rgw::sal::RadosStore*>(driver);
+    if (!store) {
+      ldpp_dout(dpp, 0) << "ERR: failed dynamic_cast to RadosStore" << dendl;
+      // TBD: should we through an exception ??
+      return;
+    }
+
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      // TBD: should we through an exception ??
+      return;
+    }
+    // generate an empty epoch with zero counters
+    int ret = set_epoch(ioctx, d_cluster_id, dpp, 0, 0);
+    if (ret != 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::failed set_epoch()! ret="
+			<< ret << "::" << cpp_strerror(ret) << dendl;
+      return;
+    }
+
+    dedup_epoch_t epoch;
+    reset(store, ioctx, &epoch, 0, 0);
+  }
+
+  //---------------------------------------------------------------------------
+  int cluster::reset(rgw::sal::RadosStore *store,
+		     librados::IoCtx &ioctx,
+		     dedup_epoch_t *p_epoch,
+		     work_shard_t num_work_shards,
+		     md5_shard_t num_md5_shards)
+  {
+    ldpp_dout(dpp, 10) << __func__ << "::REQ num_work_shards=" << num_work_shards
+		       << "::num_md5_shards=" << num_md5_shards << dendl;
+    clear();
+
+    while (true) {
+      int ret = get_epoch(ioctx, dpp, p_epoch, __func__);
+      if (ret != 0) {
+	return ret;
+      }
+      if (p_epoch->num_work_shards && p_epoch->num_md5_shards) {
+	ldpp_dout(dpp, 10) << __func__ << "::ACC num_work_shards=" << p_epoch->num_work_shards
+			   << "::num_md5_shards=" << p_epoch->num_md5_shards << dendl;
+	break;
+      }
+      else if (!num_work_shards && !num_md5_shards) {
+	ldpp_dout(dpp, 10) << __func__ << "::Init flow, no need to wait" << dendl;
+	break;
+      }
+      else {
+	ret = swap_epoch(dpp, ioctx, p_epoch, p_epoch->dedup_type,
```

Comment:
you ignore the return code. what is swap fails?

Reply:
The first line in the loop calls get_epoch() and the decision is based on the values read by it.
I don't care who set the values (all members are racing here)
The loop stops when both p_epoch->num_work_shards and 
p_epoch->num_md5_shards hold valid entries